packages/gdpr-tool/src/main.js

```diff
@@ -0,0 +1,215 @@
+import flatten from 'lodash.flatten'
+import silentLogger from './utils/silent-logger'
+import pkg from '../package.json'
+
+// todo add flow types
+export default class GDPRTool {
+  constructor(options) {
+    if (!options.apiConfig)
+      throw new Error('The constructor must be passed an `apiConfig` object')
+    this.apiConfig = options.apiConfig
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareWithExistingToken(
+          options.accessToken ? `Bearer ${options.accessToken}` : ''
+        ),
+        createAuthMiddlewareForClientCredentialsFlow(this.apiConfig),
+        createUserAgentMiddleware({
+          libraryName: pkg.name,
+          libraryVersion: pkg.version,
+        }),
+        createHttpMiddleware({ host: this.apiConfig.apiUrl }),
+      ],
+    })
+
+    this.logger = {
+      ...silentLogger,
+      ...options.logger,
+    }
+  }
+
+  getData(customerId) {
+    if (!customerId) throw Error('missing `customerId` argument')
+
+    this.logger.info('Starting fetch data')
+
+    const requestBuilder = createRequestBuilder({
+      projectKey: this.apiConfig.projectKey,
+    })
+    const customersUri = requestBuilder.customers
+      .where(`id = "${customerId}"`)
+      .perPage(500)
+      .build()
+    const ordersUri = requestBuilder.orders
+      .where(`customerId = "${customerId}"`)
+      .perPage(500)
+      .build()
+    const cartsUri = requestBuilder.carts
+      .where(`customerId = "${customerId}"`)
+      .perPage(500)
+      .build()
+    const paymentsUri = requestBuilder.payments
+      .where(`customer(id = "${customerId}")`)
+      .perPage(500)
+      .build()
+    const shoppingsListsUri = requestBuilder.shoppingLists
+      .where(`customer(id = "${customerId}")`)
+      .perPage(500)
+      .build()
+    const reviewsUri = requestBuilder.reviews
+      .where(`customer(id = "${customerId}")`)
+      .perPage(500)
+      .build()
+
+    const getUris = [
+      customersUri,
+      ordersUri,
+      cartsUri,
+      paymentsUri,
+      shoppingsListsUri,
+      reviewsUri,
+    ]
+
+    // todo change to process instead of execute
+    // in case of more than 500 results
+    return Promise.all(
+      getUris.map(uri => {
+        const request = GDPRTool.buildRequest(uri, 'GET')
+
+        return this.client.execute(request)
+      })
+    )
+      .then(async data => {
+        let results = flatten(data.map(response => response.body.results))
+        const ids = results.map(result => result.id)
+
+        if (ids.length > 0) {
+          const reference = GDPRTool.buildReference(ids)
+
+          const messagesUri = requestBuilder.messages
+            .where(reference)
+            .perPage(500)
+            .build()
+
+          const request = GDPRTool.buildRequest(messagesUri, 'GET')
+          const messages = await this.client.execute(request)
+
+          results = [...messages.body.results, ...results]
```

Comment:
I fail to understand what gets combined here? The nested but flattened response of a `Promise.all` before and what does the `messages` request return?

Reply:
Good question. `messages` returns the response from this
https://docs.commercetools.com/http-api-projects-messages.html

We build the predicate to all messages [here](https://github.com/commercetools/nodejs/blob/45d0f7902d0cd43c8e92d281eba81e22b56ccf88/packages/gdpr-tool/src/main.js#L89) since they can be linked to not only the `customerID`, but also the `orderId`, `cartId` etc etc.

So we then combine all the results(customer,cart,orders, etc) with all messages.